Use GenAI package for google

ex0ns opened the pull request:
# Description

Google deprecated the google-generativeai package and replaced it with a unified GenAI package.
This is a first attempt to make the llama index llm interface work with this packages.

We used it internally for the last week and noticed no issues so far (keep in mind that we do not use most of the feature that llama index provide, we mostly use workflow and the unified LLM interface).

One point that need to be improved/check is the usage of the kwargs. Other  packages let you override almost anything using the kwargs/llm_kwargs method, but in this case I went straight to the point and remove some of this "kwargs handling" (or even settings handling from the constructor).

## New Package?

I created a new package so I could easily test it, but we might want to add this to the existing Gemini package ? I let you decide.

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Your pull-request will likely not be merged unless it is covered by some form of impactful unit testing.

- [x] I added new unit tests to cover this change
- [ ] I believe this change is already covered by existing unit tests

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

ycd commented on llama-index-integrations/llms/llama-index-llms-genai/llama_index/llms/genai/base.py:
```diff
@@ -0,0 +1,504 @@
+        default=DEFAULT_TEMPERATURE,
+        description="The temperature to use during generation.",
+        ge=0.0,
+        le=1.0,
```
Is there a reason to limit it with 1.0?

The newer models since Gemini 1.5, the supported range is between 0.0 - 2.0.

logan-markewich commented on llama-index-integrations/llms/llama-index-llms-genai/llama_index/llms/genai/base.py:
```diff
@@ -0,0 +1,504 @@
+        api_base: Optional[str] = None,
+        transport: Optional[str] = None,
+        default_headers: Optional[Dict[str, str]] = None,
+        **generate_kwargs: Any,
```
It would be super cool to have some option to support vertexai in these options (I see the genai docs list some specific args for this)

TheMellyBee replied:
Yes please! My team needs to be able to have the VertexAI Support flag and parameters to be able to be turned on.

logan-markewich commented on llama-index-integrations/llms/llama-index-llms-genai/llama_index/llms/genai/base.py:
```diff
@@ -0,0 +1,504 @@
+        le=1.0,
+    )
+    max_tokens: int = Field(
+        default=DEFAULT_NUM_OUTPUTS,
```
Does the API like this being `None` or nah?

ex0ns replied:
It does not seem to care when being called with None, I can remove this check

logan-markewich commented on llama-index-integrations/llms/llama-index-llms-genai/llama_index/llms/genai/base.py:
```diff
@@ -0,0 +1,504 @@
+        self.model = model
+        self._client = genai.Client(**config_params)
+        self._model_meta = self._client.models.get(model=model)
+        self._is_function_call_model = True
```
I wonder if this needs to be an init arg if we support vertex

logan-markewich commented on llama-index-integrations/llms/llama-index-llms-genai/llama_index/llms/genai/base.py:
```diff
@@ -0,0 +1,504 @@
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "Gemini_LLM"
```
nit: rename

logan-markewich commented on llama-index-integrations/llms/llama-index-llms-genai/llama_index/llms/genai/base.py:
```diff
@@ -0,0 +1,504 @@
+        return gen()
+
+    @llm_chat_callback()
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
```
The sync and async methods are missing a lot of the tool calling stuff

ex0ns replied:
Isn't this done inside prepare_chat_params ? Did I miss something ?

logan-markewich replied:
So `_prepare_chat_with_tools()` prepares all the kwargs, `get_tool_calls_from_response` parses the tool calls from the response, but `chat()/achat()/etc.` need to actually attach the tool calls as well (check out the Ollama LLM for a decent example tbh, or the existing Gemini class)

logan-markewich reviewed (commented):
This seems like a good start! I'll try it out in a day or two and possibly make some changes if I run into things

ex0ns commented on llama-index-integrations/llms/llama-index-llms-genai/llama_index/llms/genai/base.py:
```diff
@@ -0,0 +1,504 @@
+            # When using `transport="rest"`, these will be sent as HTTP headers.
+            config_params["default_metadata"] = default_metadata
+
+        max_tokens = max_tokens or 2**13
```
We can

ex0ns commented:
I cleaned up a bit the options/config management, I think that a rework of `_prepare_chat_with_tools` will be required, as well as `structured_predict` (and its variants) to make sure that we are correctly propagating the configurations and options to the model, I think that I would need hand from someone that knows quite well how those methods are used and called from within LlamaIndex @logan-markewich.

I added the config for VertexAI but did not test any of them.

Another open point about configuration, is how to deal with the chat vs message setup, 

```python
    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
        kwargs_gen_config = kwargs.pop("generation_config", {})
        generation_config = {**(self._generation_config or {}), **kwargs_gen_config}
        params = {**kwargs, "generation_config": generation_config}
        next_msg, chat_kwargs = prepare_chat_params(self.model, messages, **params)
        chat = self._client.chats.create(**chat_kwargs)
        response = chat.send_message(next_msg.parts)  # type: ignore
        return chat_from_gemini_response(response)
  ```
  
  GenerationConfig here could be passed both to `chats.create` (what I'm doing right now), but also to `send_message` (that I ignored). tbh as those two methods are called next to each other, it might not even matter which one we pick. The most interesting/complex part is that the configuration is not incremental at all, this is why we need to always merge the configuration (self._generation_config) with the kwargs. The easy side is that everything is Optional in google's object so it's as easy to add a configuration, but also to remove it (i.e explicitly set it to None in the kwargs). But maybe I'm overthinking it, let me know what you think.

logan-markewich commented:
Ok, I'm pretty happy with this now -- will merge! (note the slight rename of the package)

Thanks a ton for the help!

ex0ns commented:
Thank for taking care of the remaining work for this to work, I have been super busy and was not able to find the time to finish all of that 👌

TheMellyBee commented:
Whoop!!! This is super exciting!
